### PR TITLE
check-encryption-leak: always report Overlay traffic post VinE

### PR DESCRIPTION
**Pull request by smagnani96**
**CLOSING** as motivated in the [comment below](https://github.com/cilium/cilium/pull/41884#issuecomment-3401544819)

check-encryption-leaks.bt: 
* **NativeRouting**: We always need to inspect headers to determine if a plain-text packet is a leak. Currently, we only detect:
     * pod-to-pod.
     * pod-to-pod-via-proxy
* **OverlayRouting**: Currently behaves the same as NativeRouting. However, with VinE (>= v1.18), both IPSec and WireGuard encrypt all overlay traffic. This means we could potentially flag all plain-text overlay packets as leaks, catching far more cases than just pod-to-pod or pod-to-pod-via-proxy.

The PR aims to simplify leak detection for overlay traffic, which should always be encrypted.

**TL;DR;** of ongoing discussion below https://github.com/cilium/cilium/pull/41884#discussion_r2388541995: we cannot catch XDP_TX-ed packets from our current kprobes (mainly `br_forward` and `__dev_queue_xmit`). Therefore DSR Geneve packets when using NodePort acceleration (XDP) are not observed, thus this change should be safe (i.e., all overlay traffic we match in kprobes must be encrypted).

### CODE REVIEW

File: .github/actions/bpftrace/scripts/check-encryption-leaks.bt [hunk 1]
```diff
@@ -105,7 +106,27 @@ kprobe:br_forward
         ($proto == PROTO_IPV6 && $ip6h->nexthdr == PROTO_UDP)) &&
       (bswap($udph->dest) == PORT_VXLAN || bswap($udph->dest) == PORT_GENEVE);
 
+  // Overlay packets should be always encrypted. Thus, let's log tunnel headers
+  // and shift references to the inner packet for further logging.
   if ($skb->encapsulation || $encap) {
```
Comment (julianwiedmann) [inline .github/actions/bpftrace/scripts/check-encryption-leaks.bt]: I wonder if this might give us false-positives when the node also uses those ports, for its own purposes.

As example - when we use a VXLAN overlay, we have no actual reason to complain about any unencrypted traffic on the GENEVE port.

Comment (smagnani96) [reply .github/actions/bpftrace/scripts/check-encryption-leaks.bt]: Your concern is super valid, in that case the script would fail.
But do we expect other packets to have the skb->encapsulation bit set or use the other underlay protocol port in our CI setup?
The thing is that if it's Cilium overlay traffic, we should log the leak no matter what (e.g. pod-to-pod, via proxy, etc)... Unless we use the script in some setup where we skip encryption (eg XDP remote backend with Wire guard encryption). Yeah probably here we still need to look the inner 🤔

Comment (smagnani96) [reply .github/actions/bpftrace/scripts/check-encryption-leaks.bt]: What I can do here is to add an additional parameter OVERLAY_PORT (0 in case native routing) and compare that, LGTM.

IIUC, we now don't test in CI encryption + XDP, but if we did, given #38477 we would erroneously report a leak of plain text overlay if we merged the PR as it is (with these check), right? So we should somehow instrument the bpftrace script to differentiate between such XDP-generated overlay traffic vs normal one 🤔 
cc @julianwiedmann

Comment (smagnani96) [reply .github/actions/bpftrace/scripts/check-encryption-leaks.bt]: TL;DR; we cannot catch XDP_TX-ed packets from our current kprobes (mainly `br_forward` and `__dev_queue_xmit`). Therefore DSR Geneve packets when using NodePort acceleration (XDP) are not observed, thus this change should be safe (i.e., all overlay traffic we match in kprobes must be encrypted).

Even though we don't test XDP with IPSec/WireGuard, I was stubborn enough to go and test it.
Basically, our current probes are not enough to catch XDP packets in bpftrace.
I tried using `kprobe:generic_xdp_tx` and the hook is executed, but operations are pretty much limited given `struct xdp_buff *` usages are not supported. The only sensed operation I could do was to print the pointer `$xdp` itself, but no `$xdp->data` or others.
I ended up seeing those packets in `kprobe:ip_local_deliver`, which is executed in the destination (remote) node with the chosen backend. This hook (already TC) is executed right before the kernel forwards it to the Geneve interface for decap.

To conclude, I don't see potential flakes into "Always reporting plain-text overlay as leak" in our bpftrace script.
However, this has opened my mind... I believe having a custom network traffic analyzer would give us more freedom to catch other edge cases in future, and personalize the logic even further.
(this also would pave the way to what Gray had in mind: passing a payload "I MUST BE ENCRYPTED" in CLI tests, and report a leak if our tcpdump-like sniffer sees it plaintext).

**Comment by smagnani96** [comment]
Rebased to pull in latest changes to the script and have more logs while printing leaks.
This is to investigate previous failure https://github.com/cilium/cilium/actions/runs/18460318666/job/52590104630.
From [sysdumps](https://github.com/user-attachments/files/22883100/cilium-sysdumps-ipsec-6.zip), it looks strange (overlay traffic against an external pod?)

**Comment by smagnani96** [comment]
Closing this PR, as this cannot be implemented right now as is.
By looking at `ci-ipsec-e2e` [run](https://github.com/cilium/cilium/actions/runs/18466339369/job/52609132462), the `ipsec-6` configuration (EGW) consistently reports a false alarm:

```bash
[12:55:55:806475] [0xffff888105727400] 172.20.0.3:53066 -> 172.20.0.2:8472 (proto: 17, encap: 1, skb: 0)
[12:55:55:806475] [0xffff888105727400]  ↳ 10.244.1.152:0 -> 172.20.0.5:0 (proto: 1, len: 84, ifindex: 7, netns: f0000000, srcPod: 1 (internal: 0), dstPod: 0 (internal: 0), proxy: 0 (masqueraded: 0))
[12:55:55:806475] [0xffff888105727400]   ↳ ICMP (Type: 8, Code: 0, IPFlags: .D., FragOff: 0, FragID: 16534)
```

This is EgressGateway traffic from one node (172.20.0.3) to another (172.20.0.2) for a pod (10.244.1.152) to external pod (172.20.0.5) connection. Same can be observed a bit below in the CI run for pod-to-external-service.
Thanks to Julian, who helped pointing this out, we're now misaligned between WG and IPSec:
* WG: we use `HAVE_ENCAP` ([here](https://github.com/cilium/cilium/blob/main/bpf/lib/wireguard.h#L74-L80)) to decide whether to encrypt or not. This applies to all overlay traffic, even in native routing mode. `HAVE_ENCAP` will still be true due to EGW.
* IPSec: we use `TUNNEL_MODE` ([here](https://github.com/cilium/cilium/blob/main/bpf/lib/encrypt.h#L215-L234)), which is set only with VXLAN/Geneve overlay mode.

That's the reason this issue appeared only in the IPSec-related CI run.
Therefore, in our `check-encryption-leaks.bt` we cannot simply say "all plain-text overlay is a flake" right now.
